### DISCUSSION TITLE
Test Results for Firmware

### DISCUSSION THREAD

**Discussion by jimahlstrom** [softerhardware/CWKeyer#33] [General]
I loaded firmware into the SofterHardware CWKeyer and got these results on Linux on Feb 16, 2022:

Firmware: hasak/hasak.TEENSY40@600.hex
Sound Card Name: hasak USB Audio
MIDI Device Name: hasak MIDI 1
Straight Key: 0x9000,  0 or 127 
PTT In: No message
Paddle: Same as straight key
Master Volume, Sidetone Vol, Sidetone Freq: 0xB0 and 63, 62, 6, 26; apparently NRPN 
Speed: 0xB026, 5 to 54

Firmware: TeensyWinkeyEmulator.ino.TEENSY40.hex
Sound Card Name: KeyerShield USB Audio
MIDI Device Name: KeyerShield MIDI 1
Straight Key: 0x9912,127; pause, 0x9911, 0 or 127; 0x9912, 0
PTT In: 0x9912, 0 or 127, tip and ring the same
Paddle: Same as straight key
Master Volume: 0xB907 0 to 127
Sidetone Volume: 0xB90C, 0 to 127
Sidetone Frequency: 0xB90D, 40 to 103
Speed: 0xB94A, 5 to 52

It would be nice if the keyer firmwares were more compatible. In particular, the sound and MIDI names should be the same. The pause for Winkey is still there. I know I can set it to zero with a MIDI message, but Quisk lacks the ability to send MIDI messages. And I don't see the point of NRPM messages because 14-bit values are not required. Also, I believe the standard for NRPN is 99, 98, 6, 38. See [https://en.wikipedia.org/wiki/NRPN](url).

Jim, N2ADR

**Comment by softerhardware** [comment]
Yes, I would like to see the firmware more compatible too. I think we are using 99, 98, 6 and 38 for NRPN. Where do you see a problem?

73,

Steve
kf7o

**Comment by jimahlstrom** [comment]
I thought hasak was using different numbers but that might be my error. I will retest.

**Comment by jimahlstrom** [comment]
I retested and got the following:

Firmware: hasak
NRPN sequence: 99, 98, 6, 38
Master Volume: NRPN, channel 0, control 3, value 16065 to 16383
Sidetone Volume: NRPN, channel 0, control 140, value 16065 to 16383
Sidetone Frequency: NRPN, channel 0, control 139, value 2560 to 12730
Speed: NRPN, channel 0, control 141, value 5 to 54

I am not sure of the value in the above. I used byte_6<<7 | byte_38. Are these values signed? I don't know how to interpret them.

Firmware TeensyWinkeyEmulator
Master Volume: ControlChange, channel 9, control 7, value 0 to 127
Sidetone Volume: ControlChange, channel 9, control 12, value 0 to 127
Sidetone Frequency: ControlChange, channel 9, control 13, value 40 to 103
Speed: ControlChange, channel 9, control 74, value 5 to 52

Jim, N2ADR

**Reply by dl1ycf** [reply]
a) all messages that are of interest to a SDR program should not use NRPN, since some (or most?) SDR programs just do not handle NRPNs. Therefore the CWKeyerShield library uses "plain vanilla" control changes for this. These are meant for sending information from the keyer to the SDR program. Admittedly, Frequency and Speed are *much* more important here than the two volumes.
The other thing, when using "simple plain vanilla control changes" is that the value must be in the range 0-127.

b) what you see is (by the way) MIDI channel No. 10 (MIDI channels 1 to 16 are internally coded as 0 to 15). Channel #10 is the default in the CWKeyerShield library. If you prefer another one, tell me. But one should not use "small numbers" (say, 1 to 4) since several "DJ MIDI consoles" used for VFO knobs etc. etc. have fixed (hard-wired) channel numbers in that range.

c) The range of values you see is not the entire (legal) range, it is just the range covered by the pots. Of course the protocol allows to "report" any CW speed between 1 and 127 wpm, but having that range for the pot would not allow fine tuning in the range 15-25 wpm which covers many operators.
So the ranges have been chosen as to cover the "important" range and having fine-tuning possiblities at the same time. With 5 to 52 wpm CW speed, this should cover >99 percent of the "needs", and the same for a side tone frequency 400 to 1030 Hz (the MIDI controller value is the frequency in units of 10 Hz).

d) Final note: the TeensyWinkeyEmulator does not specify these settings, these are the defaults built into the CWKeyerShield library. Suggestions for other defaults are welcome, but note ControlChange #7 for master volume follows somehow the MIDI convention, so we did "some" brainstorming on this.

**Reply by jimahlstrom** [reply]
I don't disagree with any of this, but just as a point of information the serial protocol used by the WinKeyer has a command to set the range of the speed pot. For example, the user could set the range to 15 to 25 wpm.

**Reply by dl1ycf** [reply]
The pot of the CW keyer shield works a little different from the pot controlled by the keyer. The software allows you to handle the speed pot in the keyer or in the CWKEyer shield, which is the default. In the default case, the pot controls the speed independent of the Winkey settings. However, the WinKey protocol also has a "pot position reporting" through the serial line. 

This I solved as follows: if the speed changes (either by MIDI command or by the pot), the "Winkey pot" position corresponding to that speed is calculated and reported back. The pot position is always a number between 0 and 31. If the speed is outside the range MinWPM ... MinWPM+31, then either position 0 or position 31 is reported back.

Note that in host mode, the pot normally has no effect since the "host speed" overrides. But most controlling programs look for the "pot change" messages and set the speed accordingly. So, if you turn the pot controlled b the CWKeyerShield library, a pot position is calculated therefrom and reported to the host, which in turn will change the speed. So this usually works very well, within the restriction given, namely that the speed can be controlled by several independent instances.

I work every day with the keyer, also doing contests etc., and the way I arranged things is (for me) very convenient, but if you have a suggestion for an improvement it will be considered. But these suggestions should not come from formal consideration but rather from "real QSO" or "real contesting" experience.

**Reply by jimahlstrom** [reply]
Yes, I agree that the design should be driven by the needs and expectations of real CW enthusiasts.

**Reply by recri** [reply]
Setting the default MIDI channel to 10 is fine with me.  But I do not see how it matters.  The keyer is the only device on its MIDI interface, there are no other devices on the interface that could have channel conflicts.

As for only using control changes, remember that Reid has legacy SDRs that expect to communicate volumes, tone, and speed via MIDI NOTE messages.

**Comment by recri** [comment]
The points of NRPN messages are:

1) To have more controls than will fit into the 120 midi control change
messages.
2) To have parameters which use more than 7 bit values, and more of them
than would fit into the optional 14 bit control change messages.
3) To have parameters which require more than 14 bit values.
4) To have a consistent interface rather than one that hops willy nilly
between 7 bit and 14 bit and control change and NRPN.
5) To not have a MIDI controller which tries to pun itself into the defined
MIDI control change set.
6) To have a large enough parameter set to not have to fiddle with the
design and implementation issues of packing and unpacking bit fields.

Consistency between implementations would be good.  The plan for hasak has
always been to emulate the CWKeyerShield/TeensyWinkeyEmulator control set,
but it wasn't possible until after the last release of hasak was built.

The beginNrpn method in the teensy usb_midi.h source agrees with wikipedia
on the order of control changes 99 and 98, I'll be using that now that I've
found it.

But it really doesn't matter to hasak, it just uses the current values of
6, 38, 98, and 99 when it receives 38.  As far as hasak is concerned, MIDI
is a protocol for setting and reporting the values of a note array, a
control array, and a nrpn array.  This is not strictly MIDI compliant, but
it's very easy to understand and implement.

**Comment by jimahlstrom** [comment]
I added NRPN to my software and it seems to be working. I am writing a standalone program to control the keyer so that it can be used by hams with analog radios and not just HL2 users. In the mean time it is useful for testing.

**Comment by recri** [comment]
If you're talking to hasak, then you can use https://cwkeyer.elf.org as a
second readout of the NRPN settings.  It will do a roll call of parameters
when it first connects to the keyer, to get in sync, but thereafter it
should just watch and reflect changes as they happen, though you can use it
to change settings, too.  It runs in Chrome and Opera on linux for me.

Unfortunately it doesn't display the NRPN numbers, so you'd have to consult
the hasak/hasak.JSON file in github.  The title and label strings in there
get used in cwkeyer.elf.org, along with the units, ranges, and defined
option values with their labels and titles.

If you have a JSON or other portable object format description of the
CWKeyerShield command set, that would be very helpful.

Let me know if you have comments on the information encoded into
hasak.JSON, or suggestions about information that should be there, or
questions.
